### DISCUSSION TITLE
Upgrading r DELETES all existing R packages

### DISCUSSION THREAD

**Discussion by x1o** [Homebrew/discussions#2913] [Tap maintenance and Homebrew development] [answered]
```
==> Upgrading r
  4.1.1_1 -> 4.1.2

==> Installing dependencies for r: tcl-tk
==> Installing r dependency: tcl-tk
==> Pouring tcl-tk--8.6.12_1.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/tcl-tk/8.6.12_1: 3,045 files, 51.8MB
==> Installing r
==> Pouring r--4.1.2.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/r/4.1.2: 2,216 files, 64.6MB
==> Running `brew cleanup r`...
Removing: /usr/local/Cellar/r/4.1.1_1... (36,045 files, 945.4MB)
```

This has just deleted my ENTIRE R library. No questions asks, no warnings, nothing. CLEARLY wrong and EVIL.

Whom do I ask to fix this? Could this behaviour be controlled at the cask level, or is it inherent to the brew's updating logic?

**Comment by carlocab** [comment]
Objects in the `Cellar` are designed to be ephemeral -- don't install things there. Many other users manage to.

Have you set `HOMEBREW_NO_ENV_HINTS`? Here's what shows up when I do `brew upgrade r`:
```
❯ brew upgrade r
Warning: Treating r as a formula. For the cask, use homebrew/cask/r
==> Upgrading 1 outdated package:
r 4.1.1_1 -> 4.1.2
==> Downloading https://ghcr.io/v2/homebrew/core/r/manifests/4.1.2
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/r/blobs/sha256:46663a15e9762deaad72e5fb347c85edebc836fc071f5bd21b1948d90a9311bc
Already downloaded: /Users/carlocab/Library/Caches/Homebrew/downloads/9fe40d2fee28c18fe25ca337b7c49abfc67d672184497c5af4a548616778e6dd--r--4.1.2.monterey.bottle.tar.gz
==> Upgrading r
  4.1.1_1 -> 4.1.2

==> Pouring r--4.1.2.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/r/4.1.2: 2,216 files, 64.6MB
==> Running `brew cleanup r`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
Removing: /usr/local/Cellar/r/4.1.1_1... (2,212 files, 64.4MB)
Removing: /Users/carlocab/Library/Caches/Homebrew/r--4.1.1_1... (46.1MB)
```

The env hint answers your question. You should probably follow the recommendation to see `man brew` -- you might also be interested in `HOMEBREW_NO_CLEANUP_FORMULAE`.

**Reply by x1o** [reply]
Thank you for your answer.

I don't install my packages "there" - I use whatever sensible defaults provided by the package manager. Which turn out to be not that sensible at all.

No, I have not set `HOMEBREW_NO_ENV_HINTS`. Moreover, `r` was updated during installation of a completely unrelated package (`pdfgrep`).

I guess the correct solution would be to suggest the user to point their `R_LIBS*` outside of `/usr/local/Cellar` (in big red letters, yes 🙂) upon installation of `r`.

**Reply by carlocab** [reply]
A caveat would help, but configuring R not to install libraries into the Cellar would be better. I wonder if something like this would work: https://stackoverflow.com/a/45880501

CC @imrehorvath since I think you might have thoughts here (plus this might affect your use of Homebrew R, if you're still using it)

**Comment by imrehorvath** [comment]
Hello @carlocab, @x1o,
Thanks for the info! Actually I don't use Homebrew R any more. I choose to use the official CRAN distributed R with a custom install. 

The official pkg contains 4 components and mentions that the latter 2, the Tcl/Tk and the Texinfo are optional. But in reality the R GUI app component is also optional. So I only install/update the R framework from the pkg. With this, I can use the CRAN distributed binary packages too. As IDE, I use RStudio, not the bundled R GUI app. 

For the source packages, I have the Xcode CLT (for C/C++), and the gcc (for gfortran) installed. I use a simple Makevars file to define the Fortran compiler- and lib locations and options. I install gcc with `brew install gcc`.  With this setup I get an environment which fits my needs perfectly.

To upgrade the R framework (eg. from 4.1.1 to 4.1.2) is as simple as downloading the pkg and performing the same custom install with only the R framework selected in the installer. The installer performs the upgrade as expected.

**Comment by x1o** [comment, marked as answer]
@carlocab, @imrehorvath, I've just double-checked the default installation path of R 4.1.2, and it's `/usr/local/lib/R/4.1/site-library` -- seems like the right place for me. Nothing changed in my environment and / or R config files since I've had R 4.1.1_1, so it's a bit confusing why it now uses the correct location instead of `/usr/local/Cellar/r/4.1.2/lib/R/library` 😔

Sorry for the fuss

**Reply by carlocab** [reply]
Yup, that location is fine -- Homebrew will not touch that, so anything you've installed via R shouldn't get deleted.

**Reply by carlocab** [reply]
I had another look at it, and it is indeed not clear why your original packages were not installed into `/usr/local/lib/R/4.1/site-library`. The code that makes this so is here:

https://github.com/Homebrew/homebrew-core/blob/cb5d3dd2f4581e328645d1a3fc68290835b84351/Formula/r.rb#L119-L125

This has been around for quite a while now: https://github.com/Homebrew/homebrew-core/commit/47fb4b559204f88e2775ebc8b8d8934227d27d3a (i.e. since there was an R formula in Homebrew/core)

The only possibilities I can think of is that the symlink created by the `post_install` was somehow deleted for some reason or that the `post_install` was somehow never run.

**Reply by x1o** [reply]
Yes, thank you.

A friend of mine refers to issues like this one as "swamp gas explosions".
